File: model/aether_model/frame/codec.py
"""Payload ↔ constellation-symbol codec for one (mode, layout) pair (roadmap P1-3).

    bytes ─► bits ─► CRC-24 ─► fillers ─► LDPC (BG per §7.2.2) ─► rate matching (RV) ─►
    coprime-stride interleaver ─► Gray-labelled constellation ─► QAM symbols (time-major)

and the soft inverse, with HARQ-IR combining across redundancy versions. :class:`ToneCodec` is
the same chain for the tone floor (ADR-0013), ending in four Gray-labelled bits a tone
instead of a constellation point.

The interleaver is ``π(k) = k·p mod E`` with ``p`` the integer nearest ``E/φ`` that is coprime
with ``E``: consecutive coded bits land ≈ 0.618·E apart in the time-major symbol grid, so any
burst in time (a fade) or frequency (a notch) is scattered over the whole codeword without
the structured-interleaver failure mode of parking a run of coded bits on one carrier.
"""

from __future__ import annotations

import math
from functools import cache

import numpy as np
from numpy.typing import NDArray

from aether_model.fec.nr_ldpc import RateMatcher, nr_ldpc_code
from aether_model.frame.modes import LONG, PAYLOAD_CRC, FrameLayout, Mode, ToneKind
from aether_model.phy.constellation import constellation

BitArray = NDArray[np.uint8]
FloatArray = NDArray[np.float64]
ComplexArray = NDArray[np.complex128]

_PHI = (1 + 5**0.5) / 2


def coprime_stride(e: int) -> int:
    """Stride nearest E/φ that is coprime with E (search outward from the ideal)."""
    ideal = max(1, round(e / _PHI))
    for d in range(e):
        for cand in (ideal - d, ideal + d):
            if 1 <= cand < e and math.gcd(cand, e) == 1:
                return cand
    return 1


class FrameCodec:
    def __init__(self, mode: Mode, layout: FrameLayout = LONG) -> None:
        self.mode = mode
        self.layout = layout
        self.payload_bytes = mode.payload_bytes(layout)
        self.info_bits = mode.info_bits(layout)
        self.coded_bits = mode.coded_bits(layout)
        self.bg = mode.base_graph(layout)
        self.z = mode.lifting_size(layout)
        self.code = nr_ldpc_code(self.bg, self.z)
        if self.info_bits > self.code.k:
            raise ValueError("information block does not fit the selected lifting size")
        self.constellation = constellation(mode.modulation)
        self._stride = coprime_stride(self.coded_bits)
        k = np.arange(self.coded_bits, dtype=np.int64)
        self._perm = (k * self._stride) % self.coded_bits  # coded bit k → position π(k)
        self._inv = np.empty_like(self._perm)
        self._inv[self._perm] = k

    def _rate_matcher(self, rv: int) -> RateMatcher:
        return RateMatcher(self.code, self.info_bits, self.coded_bits, rv=rv)

    # ── transmit ──────────────────────────────────────────────────────

    def encode(self, payload: bytes, rv: int = 0) -> ComplexArray:
        if len(payload) != self.payload_bytes:
            raise ValueError(
                f"payload must be exactly {self.payload_bytes} bytes, got {len(payload)}"
            )
        bits = np.unpackbits(np.frombuffer(payload, dtype=np.uint8))
        info = np.zeros(self.code.k, dtype=np.uint8)
        info[: self.info_bits] = PAYLOAD_CRC.attach(bits)
        codeword = self.code.encode(info)
        e = self._rate_matcher(rv).match(codeword)
        interleaved = np.empty_like(e)
        interleaved[self._perm] = e
        return self.constellation.map(interleaved)

    # ── receive ───────────────────────────────────────────────────────

    def decode(
        self,
        symbols: ComplexArray,
        noise_var: float | FloatArray,
        rv: int = 0,
        buffer: FloatArray | None = None,
        max_iter: int = 25,
    ) -> tuple[bytes | None, FloatArray]:
        """Returns ``(payload or None, full-codeword LLR buffer)``.

        Pass the returned buffer back as ``buffer`` with the next redundancy version to
        combine transmissions (HARQ-IR). ``None`` means the CRC failed.
        """
        y = np.asarray(symbols, dtype=np.complex128)
        if y.shape != (self.layout.qam_symbols,):
            raise ValueError(f"expected {self.layout.qam_symbols} symbols, got {y.shape}")
        llr_interleaved = self.constellation.llr(y, noise_var)
        llr_e = llr_interleaved[self._perm]
        full = self._rate_matcher(rv).recover(llr_e, buffer=buffer)
        hard, _converged, _ = self.code.decode(full, max_iter=max_iter)
        block = hard[: self.info_bits]
        if not PAYLOAD_CRC.check(block):
            return None, full
        # The all-zero word is a codeword of every linear code and its CRC is zero, so a
        # decoder fed noise (a false detection, a frame read at the wrong start) converges
        # to it and "passes". No frame of ours is all zeros — the link layer never assigns
        # session 0 and its other frames have a non-zero kind — so the block is refused.
        if not block.any():
            return None, full
        payload = np.packbits(block[: -PAYLOAD_CRC.width]).tobytes()
        return payload, full

    # ── introspection ─────────────────────────────────────────────────

    def describe(self) -> dict[str, float | int | str]:
        return {
            "mode": self.mode.name,
            "layout": self.layout.name,
            "payload_bytes": self.payload_bytes,
            "info_bits": self.info_bits,
            "coded_bits": self.coded_bits,
            "effective_rate": round(self.info_bits / self.coded_bits, 4),
            "base_graph": self.bg,
            "z": self.z,
            "fillers": self.code.k - self.info_bits,
            "interleaver_stride": self._stride,
        }


# ── the tone floor (ADR-0013) ─────────────────────────────────────────


@cache
def gray_labels(bits: int) -> NDArray[np.uint8]:
    """Row ``t``: the bits of tone ``t``'s Gray label, most significant first — neighbouring
    tones, the ones a carrier offset or a Doppler smear confuses, differ in one bit."""
    t = np.arange(1 << bits)
    g = t ^ (t >> 1)
    return ((g[:, None] >> np.arange(bits - 1, -1, -1)[None, :]) & 1).astype(np.uint8)


class ToneCodec:
    """Payload ↔ data tones for one tone-floor kind: the OFDM codec's chain — CRC-24, LDPC,
    rate matching (RV), the golden-ratio interleaver — ending in ``log2 M`` bits a tone."""

    def __init__(self, kind: ToneKind) -> None:
        self.kind = kind
        self.bg = kind.base_graph
        self.z = kind.lifting_size
        self.code = nr_ldpc_code(self.bg, self.z)
        if kind.info_bits > self.code.k:
            raise ValueError("information block does not fit the selected lifting size")
        n = kind.coded_bits
        self._perm = (np.arange(n, dtype=np.int64) * coprime_stride(n)) % n
        m = kind.data.bits_per_symbol
        self._weights = 1 << np.arange(m - 1, -1, -1)
        self._tone_of_label = np.argsort(gray_labels(m) @ self._weights)

    def _matcher(self, rv: int) -> RateMatcher:
        return RateMatcher(self.code, self.kind.info_bits, self.kind.coded_bits, rv=rv)

    def encode(self, payload: bytes, rv: int = 0) -> NDArray[np.int64]:
        """The data symbols' tones."""
        if len(payload) != self.kind.payload_bytes:
            raise ValueError(f"payload must be {self.kind.payload_bytes} bytes, got {len(payload)}")
        info = np.zeros(self.code.k, dtype=np.uint8)
        info[: self.kind.info_bits] = PAYLOAD_CRC.attach(
            np.unpackbits(np.frombuffer(payload, dtype=np.uint8))
        )
        e = self._matcher(rv).match(self.code.encode(info))
        bits = np.empty_like(e)
        bits[self._perm] = e
        label = bits.reshape(-1, self.kind.data.bits_per_symbol).astype(np.int64) @ self._weights
        return np.asarray(self._tone_of_label[label], dtype=np.int64)

    def decode(
        self,
        llr_bits: FloatArray,
        rv: int = 0,
        buffer: FloatArray | None = None,
        max_iter: int = 30,
    ) -> tuple[bytes | None, FloatArray]:
        """``(payload or None, the codeword's LLR buffer)`` from the data symbols' bit LLRs
        (positive = 0) in transmission order; the buffer combines retransmissions (HARQ-IR)
        as :meth:`FrameCodec.decode`'s does."""
        llr_e = np.asarray(llr_bits, dtype=np.float64)[self._perm]
        full = self._matcher(rv).recover(llr_e, buffer=buffer)
        hard, _converged, _ = self.code.decode(full, max_iter=max_iter)
        block = hard[: self.kind.info_bits]
        # the all-zero word passes every linear code and its CRC: refused, as FrameCodec does
        if not PAYLOAD_CRC.check(block) or not block.any():
            return None, full
        return np.packbits(block[: -PAYLOAD_CRC.width]).tobytes(), full


@cache
def tone_codec(kind: ToneKind) -> ToneCodec:
    return ToneCodec(kind)

File: core/aether-phy/src/codec.rs
//! Payload ↔ constellation-symbol codec for one (mode, layout) pair.
//!
//! ```text
//! bytes ─► bits ─► CRC-24 ─► fillers ─► LDPC ─► rate matching (RV) ─►
//! coprime-stride interleaver ─► Gray-labelled constellation ─► QAM symbols (time-major)
//! ```
//!
//! and the soft inverse, with HARQ-IR combining across redundancy versions.
//!
//! The interleaver is `π(k) = k·p mod E`, with `p` the integer nearest `E/φ` that is coprime
//! with `E`. Consecutive coded bits land about 0.618·E apart in the time-major symbol grid,
//! so a burst in time (a fade) or in frequency (a notch) is scattered over the whole
//! codeword — without the failure mode of a structured interleaver, which is to park a run
//! of coded bits on one carrier.

use aether_fec::{
    CRC24A,
    ldpc::{FecError, NrLdpcCode},
    rate_match::RateMatcher,
};

use crate::{
    constellation::{Complex, Constellation, NoiseVar},
    modes::{FrameLayout, Mode},
};

/// Golden ratio, for the interleaver stride.
const PHI: f64 = 1.618_033_988_749_895;

/// Greatest common divisor.
const fn gcd(mut a: usize, mut b: usize) -> usize {
    while b != 0 {
        let t = a % b;
        a = b;
        b = t;
    }
    a
}

/// Stride nearest `E/φ` that is coprime with `E`, searching outward from the ideal.
#[must_use]
pub fn coprime_stride(e: usize) -> usize {
    if e <= 1 {
        return 1;
    }
    let ideal = ((e as f64 / PHI).round() as usize).max(1);
    for delta in 0..e {
        for candidate in [ideal.wrapping_sub(delta), ideal + delta] {
            if candidate >= 1 && candidate < e && gcd(candidate, e) == 1 {
                return candidate;
            }
        }
    }
    1
}

/// Everything needed to turn a payload into symbols and back for one mode and layout.
#[derive(Debug, Clone)]
pub struct FrameCodec {
    mode: Mode,
    layout: FrameLayout,
    /// Payload bytes carried per frame.
    pub payload_bytes: usize,
    /// `K'`, payload plus CRC.
    pub info_bits: usize,
    /// `E`, coded bits after rate matching.
    pub coded_bits: usize,
    code: NrLdpcCode,
    constellation: Constellation,
    /// `permutation[k]` is where coded bit `k` is transmitted.
    permutation: Vec<usize>,
    /// Inverse of `permutation`.
    inverse: Vec<usize>,
    stride: usize,
}

impl FrameCodec {
    /// Build the codec.
    ///
    /// # Errors
    /// If the information block does not fit the selected lifting size.
    pub fn new(mode: Mode, layout: FrameLayout) -> Result<Self, FecError> {
        let payload_bytes = mode.payload_bytes(&layout);
        let info_bits = mode.info_bits(&layout);
        let coded_bits = mode.coded_bits(&layout);
        let code = NrLdpcCode::new(mode.base_graph(&layout), mode.lifting_size(&layout))?;
        if info_bits > code.k {
            return Err(FecError::BadLength {
                expected: code.k,
                got: info_bits,
            });
        }
        let stride = coprime_stride(coded_bits);
        let mut permutation = vec![0usize; coded_bits];
        let mut inverse = vec![0usize; coded_bits];
        for (k, slot) in permutation.iter_mut().enumerate() {
            *slot = (k * stride) % coded_bits;
        }
        for (k, &position) in permutation.iter().enumerate() {
            inverse[position] = k;
        }
        Ok(Self {
            mode,
            layout,
            payload_bytes,
            info_bits,
            coded_bits,
            code,
            constellation: Constellation::new(mode.modulation),
            permutation,
            inverse,
            stride,
        })
    }

    /// The mode this codec serves.
    #[must_use]
    pub fn mode(&self) -> Mode {
        self.mode
    }

    /// The layout this codec serves.
    #[must_use]
    pub fn layout(&self) -> FrameLayout {
        self.layout
    }

    /// The interleaver stride, for diagnostics and the specification.
    #[must_use]
    pub fn interleaver_stride(&self) -> usize {
        self.stride
    }

    /// Payload bytes → constellation symbols at the given redundancy version.
    ///
    /// # Errors
    /// If the payload is not exactly [`payload_bytes`](Self::payload_bytes) long.
    pub fn encode(&self, payload: &[u8], rv: u8) -> Result<Vec<Complex>, FecError> {
        Ok(self.constellation.map(&self.encode_bits(payload, rv)?))
    }

    /// The interleaved coded bits a payload produces, before they are mapped onto points.
    ///
    /// This is the quantity that has to match the reference model exactly; mapping bits to
    /// constellation points is a separate step with its own test.
    ///
    /// # Errors
    /// If the payload is not exactly [`payload_bytes`](Self::payload_bytes) long.
    pub fn encode_bits(&self, payload: &[u8], rv: u8) -> Result<Vec<u8>, FecError> {
        if payload.len() != self.payload_bytes {
            return Err(FecError::BadLength {
                expected: self.payload_bytes,
                got: payload.len(),
            });
        }
        let mut bits = Vec::with_capacity(payload.len() * 8);
        for &byte in payload {
            for shift in (0..8).rev() {
                bits.push((byte >> shift) & 1);
            }
        }
        let with_crc = CRC24A.attach(&bits);
        debug_assert_eq!(with_crc.len(), self.info_bits);

        let mut info = with_crc;
        info.resize(self.code.k, 0); // fillers
        let codeword = self.code.encode(&info)?;
        let matcher = RateMatcher::new(&self.code, self.info_bits, self.coded_bits, rv)?;
        let selected = matcher.match_bits(&codeword)?;

        let mut interleaved = vec![0u8; self.coded_bits];
        for (k, &bit) in selected.iter().enumerate() {
            interleaved[self.permutation[k]] = bit;
        }
        Ok(interleaved)
    }

    /// Constellation symbols → payload, or `None` if the CRC fails.
    ///
    /// `buffer` carries the full-codeword LLRs of an earlier transmission of the same block;
    /// pass the returned buffer back with the next redundancy version to combine them.
    ///
    /// # Errors
    /// If `symbols` is not the layout's slot count.
    pub fn decode(
        &self,
        symbols: &[Complex],
        noise_var: NoiseVar<'_>,
        rv: u8,
        buffer: Option<&[f64]>,
    ) -> Result<(Option<Vec<u8>>, Vec<f64>), FecError> {
        let expected = self.layout.qam_symbols();
        if symbols.len() != expected {
            return Err(FecError::BadLength {
                expected,
                got: symbols.len(),
            });
        }
        let interleaved = self.constellation.llr(symbols, noise_var);
        let mut llr_e = vec![0.0f64; self.coded_bits];
        for (k, slot) in llr_e.iter_mut().enumerate() {
            *slot = interleaved[self.permutation[k]];
        }
        let matcher = RateMatcher::new(&self.code, self.info_bits, self.coded_bits, rv)?;
        let full = matcher.recover(&llr_e, buffer)?;
        let decoded = self.code.decode(&full, 25, 0.8)?;

        let block = &decoded.bits[..self.info_bits];
        if !CRC24A.check(block) {
            return Ok((None, full));
        }
        // The all-zero word is a codeword of every linear code and its CRC is zero, so a
        // decoder fed noise (a false detection, a frame read at the wrong start) converges to
        // it and "passes". No frame of ours is all zeros — the link layer never assigns
        // session 0 and its other frames have a non-zero kind — so the block is refused.
        if block.iter().all(|&bit| bit == 0) {
            return Ok((None, full));
        }
        let payload_bits = &block[..block.len() - CRC24A.width as usize];
        let payload = payload_bits
            .chunks_exact(8)
            .map(|chunk| chunk.iter().fold(0u8, |acc, &b| (acc << 1) | (b & 1)))
            .collect();
        Ok((Some(payload), full))
    }

    /// Deinterleave order, exposed for tests and diagnostics.
    #[must_use]
    pub fn inverse_permutation(&self) -> &[usize] {
        &self.inverse
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modes::{CONTROL_MODE, LONG, MODES, SHORT};

    #[test]
    fn the_stride_is_coprime_and_near_the_golden_ratio() {
        for e in [1176usize, 2352, 3528, 4704, 7056, 420] {
            let p = coprime_stride(e);
            assert_eq!(gcd(p, e), 1, "stride {p} shares a factor with {e}");
            let ideal = e as f64 / PHI;
            assert!(
                (p as f64 - ideal).abs() < 16.0,
                "stride {p} is far from {ideal}"
            );
        }
    }

    #[test]
    fn the_interleaver_is_a_permutation() {
        let codec = FrameCodec::new(MODES[4], LONG).unwrap();
        let mut seen = vec![false; codec.coded_bits];
        for &position in &codec.permutation {
            assert!(!seen[position], "the interleaver is not injective");
            seen[position] = true;
        }
        assert!(seen.into_iter().all(|s| s));
        for (k, &position) in codec.permutation.iter().enumerate() {
            assert_eq!(codec.inverse[position], k);
        }
    }

    #[test]
    fn every_mode_round_trips_a_payload() {
        for &m in &MODES {
            let codec = FrameCodec::new(m, LONG).unwrap();
            let payload: Vec<u8> = (0..codec.payload_bytes)
                .map(|i| ((i * 37) % 256) as u8)
                .collect();
            let symbols = codec.encode(&payload, 0).unwrap();
            assert_eq!(symbols.len(), LONG.qam_symbols(), "{}", m.name());
            let (decoded, _) = codec
                .decode(&symbols, NoiseVar::Uniform(0.005), 0, None)
                .unwrap();
            assert_eq!(decoded.as_deref(), Some(&payload[..]), "{}", m.name());
        }
    }

    #[test]
    fn the_control_frame_round_trips_on_the_short_layout() {
        let codec = FrameCodec::new(CONTROL_MODE, SHORT).unwrap();
        let payload = vec![1u8, 2, 3, 4, 5, 6, 7];
        let symbols = codec.encode(&payload, 0).unwrap();
        let (decoded, _) = codec
            .decode(&symbols, NoiseVar::Uniform(0.01), 0, None)
            .unwrap();
        assert_eq!(decoded.as_deref(), Some(&payload[..]));
    }

    #[test]
    fn a_corrupted_frame_reports_failure_rather_than_garbage() {
        let codec = FrameCodec::new(MODES[10], LONG).unwrap();
        let payload: Vec<u8> = (0..codec.payload_bytes).map(|i| (i % 251) as u8).collect();
        let mut symbols = codec.encode(&payload, 0).unwrap();
        for symbol in symbols.iter_mut().step_by(2) {
            *symbol = (-symbol.1, symbol.0); // rotate a quarter turn: heavy damage
        }
        let (decoded, _) = codec
            .decode(&symbols, NoiseVar::Uniform(0.05), 0, None)
            .unwrap();
        assert!(
            decoded.is_none(),
            "the CRC must catch this rather than return wrong bytes"
        );
    }

    #[test]
    fn wrong_payload_length_is_an_error() {
        let codec = FrameCodec::new(MODES[0], LONG).unwrap();
        assert!(matches!(
            codec.encode(&[1, 2, 3], 0),
            Err(FecError::BadLength { .. })
        ));
        assert!(matches!(
            codec.decode(&[(0.0, 0.0)], NoiseVar::Uniform(1.0), 0, None),
            Err(FecError::BadLength { .. })
        ));
    }
}
